subprojects/language-native/src/main/java/org/gradle/language/nativeplatform/internal/toolchains/DefaultToolChainSelector.java

```diff
@@ -39,28 +40,48 @@ public DefaultToolChainSelector(ModelRegistry modelRegistry) {
 
     @Override
     public <T extends NativePlatform> Result<T> select(Class<T> platformType) {
-        DefaultNativePlatform targetPlatform = new DefaultNativePlatform("current");
-        NativeToolChainInternal toolChain = (NativeToolChainInternal) modelRegistry.realize("toolChains", NativeToolChainRegistryInternal.class).getForPlatform(targetPlatform);
-        PlatformToolProvider toolProvider = toolChain.select(targetPlatform);
+        DefaultNativePlatform platformRequest = new DefaultNativePlatform("current");
 
-        T t = null;
+        NativeToolChainInternal toolChain = (NativeToolChainInternal) modelRegistry.realize("toolChains", NativeToolChainRegistryInternal.class).getForPlatform(platformRequest);
+        PlatformToolProvider toolProvider = toolChain.select(platformRequest);
+
+        T targetPlatform = null;
         if (CppPlatform.class.isAssignableFrom(platformType)) {
-            t = platformType.cast(new DefaultCppPlatform("current"));
+            targetPlatform = platformType.cast(new DefaultCppPlatform("current"));
         } else if (SwiftPlatform.class.isAssignableFrom(platformType)) {
-            t = platformType.cast(new DefaultSwiftPlatform("current"));
+            if (toolProvider.isAvailable()) {
+                try {
+                    targetPlatform = platformType.cast(new DefaultSwiftPlatform("current", SwiftVersion.of(toolProvider.getCompilerMetadata().getVersion())));
+                } catch (IllegalArgumentException ex) {
+                    targetPlatform = platformType.cast(new UnknownSwiftPlatform("current"));
```

Comment:
I would probably propagate this, or at least give some different information in the failure that a user receives when querying the platform so that they can distinguish between the case where we can't determine the compiler version and the case where we don't recognise the compiler version. Or get rid of the enum so that this failure goes away.

It would also be good to not lose the information from the tool provider as to why the version could not be determined.